### PR TITLE
Tdl 18011 add missing tap tester test cases

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
**TDL-18011 Add Missing Tap Tester Test Cases**

- Added following below integration tests
  - Start Date Test
  - Pagination Test
  - Automatic Fields Test
  - All Fields Test
  - Bookmark Test
  - Discovery Test

**TDL-18068 Make replication key with automatic inclusion**
- Added function in `schema.py` to make replication key as automatic

**TDL-18458 Make pagination parameter configurable**
- Made pagination limit(page_size) configurable.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by KrisPersonal — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

File: tap_twitter_ads/schema.py [hunk 1]
```diff
@@ -73,6 +73,17 @@ def load_shared_schema_refs():
     return shared_schema_refs
 
 
+def make_replication_key_automatic(mdata, schema, replication_keys):
```
Comment (dbshah1212) [inline tap_twitter_ads/schema.py:76]: Add function comment

Comment (prijendev) [reply tap_twitter_ads/schema.py:76]: Added comment.

File: tests/test_twitter_ads_bookmark.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                    self.assertIsNone(first_bookmark_value)
+                    self.assertIsNone(first_bookmark_value)
+
+                    # Verify the number of records in the second sync is the same as the first
+
+                    # `targeting_criteria` is child stremas of parent stream `line_items` and tickets is incremental streams
+                    # Child stream also behave like incremental streams but does not save it's own state. So, it don't 
+                    # have same no of record on second sync and first sync.
+                    if stream in ['targeting_criteria']:
+                        continue
```
Comment (kspeer825) [inline tests/test_twitter_ads_bookmark.py:170]: Is it possible to add some assertions specific to this child stream? My assumption is that all `target_criteria` records will be replicated that correspond to a given parent object. As the comment says the parent stream is incremental. I think it should be possible to verify consistency of this child stream across the two syncs in this test. On the initial sync we'll have all parent records and all child records for each parent. On the second sync we should have all child records for each parent record with a replication key value greater or equal to the previous bookmark. This should be possible to verify if the parent primary key is a foreign key on the child record.

**Review by kspeer825 — CHANGES_REQUESTED** [review]
Left some comments. The suite looks pretty good as a whole.

File: tests/base.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+            'access_token': os.getenv('TAP_TWITTER_ADS_ACCESS_TOKEN'),
+            'access_token_secret': os.getenv('TAP_TWITTER_ADS_ACCESS_TOKEN_SECRET')
+        }
+
+    def get_properties(self, original: bool = True):
+        """Configuration properties required for the tap."""
+
+        return_value = {
+            "account_ids": os.getenv("TAP_TWITTER_ADS_ACCOUNT_IDS"),
+            "attribution_window": os.getenv("TAP_TWITTER_ADS_ATTRIBUTION_WINDOW") or "14",
```
[the remark this answers is not in the document]
Comment (prijendev) [reply tests/base.py]: Yes, we have these environment variables. Removed fallbacks.

File: tests/base.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+                self.REPLICATION_METHOD: self.FULL_TABLE,
+                self.OBEYS_START_DATE: True
+            },
+            "targeting_interests": targeting_endpoint_metadata,
+            "targeting_languages": targeting_endpoint_metadata,
+            "targeting_locations": targeting_endpoint_metadata,
+            "targeting_network_operators": targeting_endpoint_metadata,
+            "targeting_platform_versions": targeting_endpoint_metadata,
+            "targeting_platforms": targeting_endpoint_metadata,
+            "targeting_tv_markets":  {
```
[the remark this answers is not in the document]
Comment (prijendev) [reply tests/base.py:167]: Added OBEYS_START_DATE.

File: tests/test_twitter_ads_all_fields.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                    expected_all_keys), msg='{} is not in "expected_all_keys"'.format(expected_automatic_keys-expected_all_keys))
+
+                messages = synced_records.get(stream)
+                # collect actual values
+                actual_all_keys = set()
+                for message in messages['messages']:
+                    if message['action'] == 'upsert':
+                        actual_all_keys.update(message['data'].keys())
+                    
+                # As we can't generate following fields by twitter-ads APIs and UI, so removed it form expectation list.
```
[the remark this answers is not in the document]
Comment (prijendev) [reply tests/test_twitter_ads_all_fields.py]: Updated as per suggestions.

File: tests/test_twitter_ads_bookmark.py (hunk 2 above)
Comment (prijendev) [reply tests/test_twitter_ads_bookmark.py]: @kspeer825 We have made the child stream an independent stream from the parent stream [PR#18](https://github.com/singer-io/tap-twitter-ads/pull/18). So, we have written a child-specific assertion into it. Actually, during the sync of a child, we are retrieving parents based on the bookmark stored for the child stream. Then, based on the primary key of the parent record, we retrieve the child's records. We are not storing these parent records in the target. So, the scenario that you suggested is difficult to cover here. Instead, we are verifying that the `foreign_keys` list from the second sync must be a subset of the `foreign_keys` list from the first sync.

Comment (kspeer825) [reply tests/test_twitter_ads_bookmark.py]: Got it thanks.

**Review by kspeer825 — APPROVED** [review]

**Comment by prijendev** [comment]
> @prijendev CCi job I reran failed for start date. Please look into the failure.

CCI job is passing now.

**Review by RushT007 — APPROVED** [review]

**Pull request merged**